### PR TITLE
Improve `Date.parse`

### PR DESCRIPTION
**Pull request by chqrlie**
Improve Date.parse

- rewrite Date.parse() with separate parsers for ISO and other formats
- return `NaN` for out of bounds field values as specified
- accept up to 9 decimals for millisecond fraction but truncate at 3
- accept many more alternative date/time formats
- add test cases in tests/test_builtin.js
- fix most v8 test cases

### CODE REVIEW

File: repl.js
Comment (bnoordhuis) [inline repl.js]: Hm, in my never-ceasing purge of everything and anything non-standard, this method had so far escaped my notice.

File: quickjs.c
Comment (bnoordhuis) [inline quickjs.c:47268]: Is U+0020 the only blank by spec or are e.g. U+2000 and/or U+200A also allowed? I couldn't really find anything about that in (or around) https://tc39.es/ecma262/#sec-time-zone-offset-strings, only in the lemma on Date.prototype.toString and that only mentions ASCII space.

It might be implementation-defined. V8 allows them, spidermonkey doesn't seem to.

File: quickjs.c
Comment (bnoordhuis) [inline quickjs.c:47320]:
```suggestion
            mul /= 10;
            ms += (c - '0') * mul;
```
It's a little too easy to read as `mul / 10` and overlook the assignment.

File: quickjs.c
Comment (bnoordhuis) [inline quickjs.c:47338]: Per https://tc39.es/ecma262/#sec-time-zone-offset-strings, U+2212 (MINUS SIGN) is also acceptable. I'm okay with leaving that out for now but consider adding a TODO.

File: quickjs.c
Comment (bnoordhuis) [inline quickjs.c:47364]: FWIW, it's not per spec but V8 also accepts lowercase 'z'. OTOH, spidermonkey doesn't (results in a NaN date.)

File: quickjs.c
Comment (bnoordhuis) [inline quickjs.c:47373]:
```suggestion
    return c >= 'a' && c <= 'z' ? c - 'a' + 'A' : c;
```
Else I don't understand exactly what this does. :-)

File: quickjs.c
Comment (bnoordhuis) [inline quickjs.c:47414]:
```suggestion
static BOOL js_date_parse_isostring(const uint8_t *sp, int (*fields)[9], BOOL *is_local) {
```
That way callers _must_ pass in an exactly nine-element array (and you have to dereference them here as `(*fields)[i]`)

Another alternative is C99-style `int fields[static 9]` syntax. Then callers must pass in an at least nine-element array.

File: quickjs.c
Comment (bnoordhuis) [inline quickjs.c:47425]: Not 100% sure but the comment about U+2212 probably applies to this function as well.

File: quickjs.c
Comment (bnoordhuis) [inline quickjs.c:47451]: Same as my other comment: V8 accepts lowercase 't', spidermonkey doesn't.

File: quickjs.c
Comment (bnoordhuis) [inline quickjs.c]:
```suggestion
        } else if (c == '(') {  /* skip parenthesized phrase */
```
Also, I feel slightly offended that this list leaves out CET and CEST :-)

File: quickjs.c
Comment (bnoordhuis) [inline quickjs.c]:
```suggestion
        } else if (c == ')') {
```

**Review by bnoordhuis — COMMENTED** [review]
Mostly LGTM but I believe the spec allows for some implementation-dependentness and we should decide whether we want to align with V8 (laxer) or SM (stricter.)

Although I don't generally subscribe to Postel's maxim, I'm inclined to side with V8 here because date strings are often entered by humans, not machines.

File: repl.js
Comment (chqrlie) [reply repl.js]: How about a `toSource`, which albeit non-standard is available on some systems and less surprising.  For the `Date` object output, I suggest ```std.puts(`Date "${a.toGMTString()}"`);```

File: quickjs.c
Comment (chqrlie) [reply quickjs.c:47268]: I do support the Unicode minus `U+2212` because the spec mentions it, but I had never seen one before.  Supporting other whitespace characters is possible but of limited value.

File: quickjs.c
Comment (chqrlie) [reply quickjs.c:47320]: Yes, this is a left over from quick and dirty code written in a hurry to try various approaches... not a good example in production code.  I changed the code to reduce stalls:
```
int ms = 100;
[...]
            ms += (c - '0') * mul;
            mul /= 10;
```

File: quickjs.c
Comment (chqrlie) [reply quickjs.c:47338]: The string is converted to ASCII before calling the parse functions and `U+2212` has been converted to `-` already.  This introduces a limitation on the string length (only the first 127 characters participate in the conversion) but simplifies and accelerates the parsers.

File: quickjs.c
Comment (chqrlie) [reply quickjs.c:47364]: For consistency with other abbreviations handled by `string_match`, I will accept `z` as well.

File: quickjs.c
Comment (chqrlie) [reply quickjs.c:47373]: Good catch! this passed the tests because `upper_ascii` was called on both sides of `==` for abbreviations containing lowercase letters and no tests used a different combination of cases (as in `JAN`, `jan` or `jaN`).  Fixed.

File: quickjs.c
Comment (chqrlie) [reply quickjs.c:47414]: Good suggestion.  I don't like the C99 syntax too much.  Does MSAN or ASAN catch out of bound accesses for arrays passed as pointers (without a length specified by either syntax)?

File: quickjs.c
Comment (chqrlie) [reply quickjs.c:47425]: Actually the `U+2212` is only specified for the Time Zone Offset String Format.  The string conversion adds support for this as a side effect.  Note also that the `ParseTimeZoneOffsetString` abstract operation supports longer strings with seconds and nanoseconds which I did not implement.

File: quickjs.c
Comment (chqrlie) [reply quickjs.c:47451]: `T` and `Z` are specified in uppercase for the ISO date/time format. Matching these in lowercase too does not seem necessary, but could become a side effect of the string conversion if I were to uppercase the string unconditionally. I am not sure it is worth it.

File: quickjs.c
Comment (chqrlie) [reply quickjs.c]: This is a matter of style... I like to align the `if` tests at the indentation level.  I will check if the rest of the code is consistent with that... `else if` on the same line wins 8 to 1, yet for long lists of chained tests, I find it more readable to break the line between the `else` and the `if`.

I would gladly add more timezone abbreviations and use a table.  Do V8 or SM support other ones?
I found this source: https://en.wikipedia.org/wiki/List_of_time_zone_abbreviations but it is long and some of the abbreviations are ambiguous.

This site also has a trove of useful information: https://www.timeanddate.com/time/zones/

I wrote a more general parser and added the European zones.

File: repl.js
Comment (bnoordhuis) [reply repl.js]: Why not format the string here, in place? I'm not a huge fan of swapping one non-standard method for another. toSource() was a Mozilla-ism, as far as I'm aware, and even they removed it.

edit: doesn't need to hold up this pull request since __quote is a pre-existing thing.

File: quickjs.c
Comment (bnoordhuis) [reply quickjs.c:47414]: Most of the time. Not 100% of the time though.

File: quickjs.c
Comment (bnoordhuis) [reply quickjs.c:47451]: I'd personally lean on the side of acceptance and forgiveness here but you're not wrong it's per spec so :shrug:

File: repl.js
Comment (chqrlie) [reply repl.js]: > Why not format the string here, in place?

I used ```std.puts(`Date "${a.toGMTString()}"`);``` in the latest commit.  Is this what you meant?
How to you produce the source for a string reliably without some kind of `toSource()` method?
I guess I could write it in JS and add it in `repl.js` or another package loaded by `repl.js`.

File: quickjs.c
Comment (bnoordhuis) [reply quickjs.c]: > Do V8 or SM support other ones?

Both SM's and V8's parser go well above and beyond the minimum functionality the spec mandates.

They both know how to parse e.g. `(Central European Standard Time)` (and in a case-insensitive manner); quite possibly they lean into ICU for the actual string parsing.

File: quickjs.c
Comment (chqrlie) [reply quickjs.c:47414]: I added a macro to encapsulate the C99 syntax and used that. It is less surprising to casual readers than the array pointer trick, and also allows for larger arrays if necessary, as is the case for `get_date_fields()` / `set_date_fields()` combinations.

File: repl.js
Comment (bnoordhuis) [reply repl.js]: That's what I mean, doing it inside repl.js.

File: quickjs.c
Comment (chqrlie) [reply quickjs.c:47414]: I had to disable the C99 syntax for MSVC that does not support it even though it does define `__STDC_VERSION__`.

File: v8.js
Comment (bnoordhuis) [inline v8.js]:
```suggestion
    let env = {}, envstr = ""
    for (let s of source.matchAll(/environment variables:(.+)/ig)) {
        for (let m of s[1].matchAll(/\s*([\S]+)=([\S]+)/g)) {
            env[m[1]] = m[2]
            envstr += ` ${m[1]}=${m[2]}`
        }
    }
    print(`=== ${file}${envstr}`)
```
No semicolons in my beautiful pristine JS code, please :-)

**Review by bnoordhuis — APPROVED** [review]

File: v8.js
Comment (chqrlie) [reply v8.js]: OK for consistency, but not my style :)

**Comment by saghul** [comment]
@chqrlie is this ready to go?

I'm planning on cutting a 0.4.0 release once this lands.

**Comment by chqrlie** [comment]
Yes.  just waiting for Ben's final approval

**Comment by chqrlie** [comment]
> @chqrlie is this ready to go?
> 
> I'm planning on cutting a 0.4.0 release once this lands.

Yes. just waiting for Ben's final approval

File: repl.js
Comment (chqrlie) [reply repl.js]: I implemented it in Javascript, then as I proceeded to remove `String.prototype.__quote`, I realized `JSON.stringify` does the job.  Patch submitted.

File: quickjs.c
Comment (bnoordhuis) [reply quickjs.c:47414]: The `int (*fields)[9]` syntax works in msvc; it's a c89-ism that we use quite extensively in libuv.

**Review by bnoordhuis — APPROVED** [review]
I only briefly skimmed it this time because, well, it's 10 AM on a Sunday morning and I want to go outside and play with the kids, but LGTM :)

File: quickjs.c
Comment (chqrlie) [reply quickjs.c:47414]: > The `int (*fields)[9]` syntax works in msvc; it's a c89-ism that we use quite extensively in libuv.

Of course, but this syntax would force the length to be `9` whereas `get_date_fields` needs at least `9` and  `set_date_fields` at least `7`.  It is not a problem if MSVC cannot verify the minimum length as other compilers already do for 99,9% of the code.

**Pull request merged**
